### DISCUSSION TITLE
MQTT commands and device templates

### DISCUSSION THREAD

**Discussion by maddhin** [volker76/tuya_mqtt.net#13] [Q&A] [answered]
Hi, 

I got tuya_mqtt.net successfully integrated in my FHEM system. Or to be precise: I manage to get the readings in but I seem not be able to control the tuya devices via MQTT. 

Here I have the question, how the topic needs to be structured in order to successfully send a command to tuya_mqtt.net and the device. I could not find any documentation. 

Secondly, the topics coming in are a bit cryptic. DS20/DS21/DSXX - is there a list somewhere which function/info is behind these codes? E.g. DS20 = state of power / switch, TS = time stamp, ... Such a list would be super helpful in order to have meaningful readings in the home automation server. 
And going a step further: is there maybe a way to implement templates? Say a template for tuya light bulbs, etc. which then adjusts the readings names. 

Sorry if this is a basic/stupid question but I searched the web and managed not to find anything useful in this regard. 

Cheers and a great weekend!

**Comment by volker76** [comment]
First, there is no kind of templates to translate the raw data to a kind of device type. The data to the MQTT broker is the pure DPx where the 'x' is a number 1..128 . each TUYA device can publish any number of datapoints with any meaning. This needs to be figured out by trial & error. or when having access to the cloud api - the descriptions can be read there.

**Comment by volker76** [comment, marked as answer]
About controlling the individual datapoints, if they are controllable is device dependent. you can try to write, but if the device does not like be controlled at that datapoint - nothing can be done against that.

example, you have a device publishing MQTT items to `tuya/bfcac4e409bb624cc7w9hq/DP1`    by publishing data to the item `tuya/bfcac4e409bb624cc7w9hq/DP1/command`   values can be set and written.

note: DP1 exists on all devices and this also is the controllable input.

... posting a logfile from tuya_mqtt.net may be helpful to discuss any device specific options

**Comment by volker76** [comment]
as an example: a device **bfcac4e409bb624cc7w9hq** is connected and publishes the following DP items
![Screenshot 2023-11-04 235145](https://github.com/volker76/tuya_mqtt.net/assets/20932736/3eac5a03-6cb3-40cc-bca4-f8f60b6ade51)

to identify what's what meaning - you can login to the [TUYA IOT Explorer](https://eu.iot.tuya.com/cloud/explorer)
and query device properties.

![Screenshot 2023-11-04 235916](https://github.com/volker76/tuya_mqtt.net/assets/20932736/98a7a8ea-0ed0-442f-a45f-0c7962a8ac28)

The json may give some hints what the item meaning could be.
```
{
  "result": {
    "properties": [
      {
        "code": "switch_1",
        "custom_name": "",
        "dp_id": 1,
        "time": 1699138476890,
        "value": true
      },
      {
        "code": "countdown_1",
        "custom_name": "",
        "dp_id": 9,
        "time": 1699138476890,
        "value": 0
      },
      {
        "code": "add_ele",
        "custom_name": "",
        "dp_id": 17,
        "time": 1696835079189,
        "value": 2
      },
      {
        "code": "cur_current",
        "custom_name": "",
        "dp_id": 18,
        "time": 1699138476890,
        "value": 0
      },
      {
        "code": "cur_power",
        "custom_name": "",
        "dp_id": 19,
        "time": 1699138476890,
        "value": 0
      },
      {
        "code": "cur_voltage",
        "custom_name": "",
        "dp_id": 20,
        "time": 1699138476890,
        "value": 2364
      },
      {
        "code": "test_bit",
        "custom_name": "",
        "dp_id": 21,
        "time": 1699138476890,
        "value": 1
      },
      {
        "code": "voltage_coe",
        "custom_name": "",
        "dp_id": 22,
        "time": 1699138476890,
        "value": 647
      },
      {
        "code": "electric_coe",
        "custom_name": "",
        "dp_id": 23,
        "time": 1699138476890,
        "value": 32612
      },
      {
        "code": "power_coe",
        "custom_name": "",
        "dp_id": 24,
        "time": 1699138476890,
        "value": 18833
      },
      {
        "code": "electricity_coe",
        "custom_name": "",
        "dp_id": 25,
        "time": 1699138476890,
        "value": 1108
      },
      {
        "code": "fault",
        "custom_name": "",
        "dp_id": 26,
        "time": 1699138476890,
        "value": 0
      }
    ]
  },
  "success": true,
  "t": 1699138486989,
  "tid": "26e4ae7e7b6511eeab9046acc027565a"
}
```

**Comment by volker76** [comment]
added [Wiki page](https://github.com/volker76/tuya_mqtt.net/wiki/Tuya-DEVICE-MQTT-appearance-&--control)

**Reply by maddhin** [reply]
@volker76 thank you for this info!

I tried all kinds of combinations of topics to control the devices but just adding command (no capital C!) I have not tried and now it works. 

In combination with "query device properties" in the Tuya developer IOT Platform, things a pretty clear and everybody can adjust to his device. Could you please add this info to the new wiki page, too? Then everything should be clear. 

And just a small correction: my device (cheap light bulb) does NOT have a DP1 but is controlled on DP20 as main switch. E.g. DP1 doesn't seem to be universal.  

Thanks a lot for your quick reaction! Much appreciated! And thanks for developing this "bridge" - I have been trying to use fhempy in combination with FHEM-docker and can not get it to work. Your solution is so much better as it is independent from the home automation software used. THANK YOU!

**Reply by volker76** [reply]
can you name the product or provide a link where to get it from?

**Reply by maddhin** [reply]
sorry, I just see this. I'm not 100% sure but it should be a Bewahly brand Smart Light Bulb 800 lumen - normal light bulb with E27. I got it from Amazon.de.
